### PR TITLE
Add a new metatasks controller and json serialization

### PR DESCRIPTION
**Pull request by steveburkett**
Metatasks are lightweight versions of tutoring tasks.  Used to load first more quickly, then when clicked, the full task would load thru the regular /api/tasks API.

https://app.zenhub.com/workspaces/openstax-tutor-560aef617a763234615d6a0f/issues/openstax/tutor-server/1849

### CODE REVIEW

File: app/controllers/api/v1/metatasks_controller.rb [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    ScoutHelper.ignore!(0.8)
+    standard_read(Research::ModifiedTaskForDisplay[task: @task], Api::V1::MetataskRepresenter)
+  end
+
+  api :PUT, '/tasks/:id/accept_late_work', 'Accept late work in the task score'
+  description <<-EOS
+    Changes the Task so that it is marked as late work being accepted, and changes
+    the `score` that it computes.  Does not change the underlying counts in any way.
+  EOS
+  def accept_late_work
```
Comment (steveburkett) [inline app/controllers/api/v1/metatasks_controller.rb]: @nathanstitt e.g., what to do here for these member actions

File: app/representers/api/v1/metatasks/task_step_representer.rb [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+               required: true,
+               type: 'boolean',
+               description: "Whether or not this step is complete"
+             },
+             if: NOT_FEEDBACK_ONLY
+
+    alias_method :to_hash_without_cache, :to_hash
+
+    def to_hash(*args)
+      Rails.cache.fetch(represented.cache_key, expires_in: NEVER_EXPIRES) { super }
```
Comment (steveburkett) [inline app/representers/api/v1/metatasks/task_step_representer.rb]: @nathanstitt and here...i think we need to improve this cache_key to differ between meta tasks and regular tasks.

File: app/representers/api/v1/metatasks/tasked_exercise_representer.rb [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+               description: "The title of this Exercise"
+             },
+             if: NOT_FEEDBACK_ONLY
+
+    property :content,
+             type: String,
+             writeable: false,
+             readable: true,
+             as: :content_preview,
+             getter: ->(*) { self.content.truncate(35, separator: ' ', omission: '... (preview)') },
```
Comment (steveburkett) [inline app/representers/api/v1/metatasks/tasked_exercise_representer.rb]: @nathanstitt and here...what to set metatask/step content to.

File: app/controllers/api/v1/metatasks_controller.rb (hunk 1 above)
Comment (nathanstitt) [reply app/controllers/api/v1/metatasks_controller.rb]: those are used by the teacher to accept (or reject in the corresponding method), work the student has performed after the assignment is due.

File: app/representers/api/v1/metatasks/task_step_representer.rb (hunk 2 above)
Comment (nathanstitt) [reply app/representers/api/v1/metatasks/task_step_representer.rb]: Actually we want to remove the cache altogether for the new actions.  The cache was just a bandaid on the problem of too much data being generate.

It's caused issues with the Research::ModifiedTaskForDisplay calls because they sometimes need to modify deep properties of the tasks, which the cache doesn't catch.

File: app/representers/api/v1/metatasks/tasked_exercise_representer.rb (hunk 3 above)
Comment (nathanstitt) [reply app/representers/api/v1/metatasks/tasked_exercise_representer.rb]: yes, this is going to be a tricky part of the code.  We need to come up with per-step type rules for generating the previews, and then use a lookup table or some such to run that code for each type.

These previews are used to populate what we call "milestones".  that's a slide down pane that shows each step and its status when you do a reading assignment.  It's triggered by clicking the grid icon on the navbar when you're in a reading :
image


this is roughly what I think we'll need for each step type:
 * exercise: the `stem_html` of the first question
 * external: the title and description of the task itself
 * event: does not have steps so no need for preview
 * reading: currently we're parsing the html of the step and extracting the step title using the logic here: https://github.com/openstax/tutor-js/blob/master/tutor/src/flux/step-title.js#L117-L134
   * I find that pretty hard to follow and it's always fun when we need to tweak it, but I think basically what it does is look for https://github.com/openstax/tutor-js/blob/master/tutor/src/flux/step-title.js#L36-L40 and then read it's text.
  * hopefully something like this is clearer to express using nokogiri or Oga (pick your poison) I'd thought we used nokogiri already but don't see it in the Gemfile

**Comment by steveburkett** [comment]
https://github.com/openstax/tutor-server/pull/1849

File: app/representers/api/v1/metatasks/task_step_representer.rb (hunk 2 above)
Comment (steveburkett) [reply app/representers/api/v1/metatasks/task_step_representer.rb]: @nathanstitt good call. i like the idea of cleaning up the "hacks" out of the representers and making them just pure clean json DSL.

File: app/representers/api/v1/metatasks/tasked_exercise_representer.rb (hunk 3 above)
Comment (steveburkett) [reply app/representers/api/v1/metatasks/tasked_exercise_representer.rb]: @nathanstitt how about the "interactive" step type?

File: app/representers/api/v1/metatask_representer.rb [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+             writeable: false,
+             readable: true,
+             getter: ->(*) { DateTimeUtilities.to_api_s(last_worked_at) },
+             schema_info: {
+               description: "When the metatask was last worked (nil means not yet worked)"
+             }
+
+    collection :metatask_steps,
+               as: :metatask_steps,
+               writeable: false,
```
Comment (nathanstitt) [inline app/representers/api/v1/metatask_representer.rb]: the `:as` isn't needed if the name is the same as the collection.  

Can we rename it to be just `steps`?  The front-end will map this to an association name and it'll be a bit clunky to have it be `metatask_steps`

File: app/representers/api/v1/metatasks/task_step_representer.rb [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+             writeable: false,
+             readable: true,
+             getter: ->(*) { task_step.completed? },
+             schema_info: {
+               required: true,
+               type: 'boolean',
+               description: "Whether or not this step is complete"
+             },
+             if: NOT_FEEDBACK_ONLY
+  end
```
Comment (nathanstitt) [inline app/representers/api/v1/metatasks/task_step_representer.rb]: We can take out the lambda `if: NOT_FEEDBACK_ONLY` lines on all the blocks , they'll always be true here

File: app/representers/api/v1/metatasks/task_step_representer.rb [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+             type: String,
+             writeable: false,
+             readable: true,
+             getter: ->(*) { task_step.tasks_task_id },
+             schema_info: {
+                 required: true,
+                 description: "The id of the Task"
+             },
+             if: NOT_FEEDBACK_ONLY
+
```
Comment (nathanstitt) [inline app/representers/api/v1/metatasks/task_step_representer.rb]: we don't need the task_id since the steps will be nested under the task

File: app/representers/api/v1/metatasks/tasked_exercise_representer.rb [hunk 7]
```diff
@@ -0,0 +1,34 @@
+module Api::V1::Metatasks
```
Comment (nathanstitt) [inline app/representers/api/v1/metatasks/tasked_exercise_representer.rb]: we don't need the url or title on any of the step types below.

**Comment by nathanstitt** [comment]
Thanks Steve!
